lib/elixir_analyzer/submission.ex

```diff
@@ -8,130 +8,129 @@ defmodule ElixirAnalyzer.Submission do
-          final: boolean
+          analysis_module: atom()
         }
 
-  def to_json(r = %__MODULE__{}) do
-    Jason.encode!(%{status: @status_map[r.status], comments: r.comments})
+  def to_json(submission = %__MODULE__{}) do
+    Jason.encode!(%{summary: get_summary(submission), comments: submission.comments})
   end
 
-  def approve(r = %__MODULE__{status: nil}), do: %{r | status: :approve}
-  def approve(r), do: r
-
-  def disapprove(r = %__MODULE__{status: s}) when s in [nil, :approve],
-    do: %{r | status: :disapprove}
+  @doc false
+  def halt(submission = %__MODULE__{}) do
+    %{submission | halted: true}
+  end
 
-  def disapprove(r), do: r
+  def set_halt_reason(submission = %__MODULE__{}, reason) when is_binary(reason) do
+    %{submission | halt_reason: reason}
+  end
 
-  def refer(r = %__MODULE__{}), do: %{r | status: :refer}
+  @doc false
+  def set_analyzed(submission = %__MODULE__{}, value) when is_boolean(value) do
+    %{submission | analyzed: value}
+  end
 
-  def finalize(r = %__MODULE__{status: nil, final: false}),
-    do: %{r | status: :refer} |> finalize()
+  @doc false
+  def append_comment(submission = %__MODULE__{}, meta) when is_map(meta) do
+    comment =
+      Enum.filter(meta, fn
+        {key, _} when key in [:comment, :type, :params] -> true
+        _ -> false
+      end)
+      |> Enum.into(%{})
 
-  def finalize(r = %__MODULE__{final: false}) do
-    Map.put(r, :final, true)
+    %{submission | comments: submission.comments ++ [comment]}
   end
 
-  def prepend_comment(r = %__MODULE__{}, comment) when is_binary(comment) do
-    do_prepend_comment(r, comment)
+  defp get_summary(submission = %__MODULE__{halted: true, comments: comments})
+       when length(comments) == 0 do
+    case submission.halt_reason do
+      nil -> "Analysis was halted."
+      _ -> "Analysis was halted. #{submission.halt_reason}"
+    end
   end
 
-  def prepend_comment(r = %__MODULE__{}, {comment}) when is_binary(comment) do
-    do_prepend_comment(r, %{"comment" => comment})
-  end
+  @comment_types ~w{celebratory essential actionable informative}a
+  @default_type_frequencies @comment_types |> Enum.map(&{&1, 0}) |> Enum.into(%{})
+
+  defp get_summary(%__MODULE__{comments: comments}) do
+    type_frequencies_in_comments = Enum.frequencies_by(comments, fn comment -> comment.type end)
+
+    summary =
+      @default_type_frequencies
+      |> Map.merge(type_frequencies_in_comments)
+      |> build_summary_response()
 
-  def prepend_comment(r = %__MODULE__{}, {comment, params})
-      when is_binary(comment) and is_map(params) do
-    params
-    |> Map.keys()
-    |> Enum.all?(fn key -> Kernel.is_atom(key) or Kernel.is_binary(key) end)
-    |> case do
-      false ->
-        raise ArgumentError, "key type must be atom or binary"
-
-      true ->
-        do_prepend_comment(r, %{"comment" => comment, "params" => params})
+    case summary do
+      nil -> "Submission analyzed. No automated suggestions. Great work! 🚀"
```

Comment:
I'm not sure if we should say here "great work!". It could be that somebody submitted such a weird solution that we didn't foresee it in the analyzer :D. Maybe we could be neutral?
```suggestion
      nil -> "Submission analyzed. No automated suggestions found."
```